moveit_ros/moveit_servo/src/utils.cpp

```diff
@@ -0,0 +1,319 @@
+  Eigen::VectorXd next_joint_angles;
+  current_state->copyJointGroupPositions(joint_model_group, next_joint_angles);
+  next_joint_angles += pseudo_inverse * delta_x;
+
+  // Compute the Jacobian SVD for the new robot state.
+  current_state->setJointGroupPositions(joint_model_group, next_joint_angles);
+  Eigen::JacobiSVD<Eigen::MatrixXd> next_svd = Eigen::JacobiSVD<Eigen::MatrixXd>(
+      current_state->getJacobian(joint_model_group), Eigen::ComputeThinU | Eigen::ComputeThinV);
+
+  // Compute condition number for the new Jacobian.
+  const double next_condition_number = next_svd.singularValues()(0) / next_svd.singularValues()(dims - 1);
+
+  // If the condition number has increased, we are moving towards singularity and the direction of the
+  // vector_towards_singularity is correct. If the condition number has decreased, it means the sign of
+  // vector_towards_singularity needs to be flipped.
+  if (next_condition_number <= current_condition_number)
+  {
+    vector_towards_singularity *= -1;
+  }
+
+  // Double check the direction using dot product.
+  const bool moving_towards_singularity = vector_towards_singularity.dot(target_delta_x) > 0;
+
+  // Compute upper condition variable threshold based on if we are moving towards or away from singularity.
+  // See https://github.com/ros-planning/moveit2/pull/620#issuecomment-1201418258 for visual explanation.
+  double upper_threshold;
+  if (moving_towards_singularity)
+  {
+    upper_threshold = hard_stop_singularity_threshold;
+  }
+  else
+  {
+    const double threshold_size = (hard_stop_singularity_threshold - lower_singularity_threshold);
+    upper_threshold = lower_singularity_threshold + (threshold_size * leaving_singularity_threshold_multiplier);
+  }
+
+  // Compute the scale based on the current condition number.
+  double velocity_scale = 1.0;
+  const bool is_above_lower_limit = current_condition_number > lower_singularity_threshold;
+  const bool is_below_hard_stop_limit = current_condition_number < hard_stop_singularity_threshold;
+  if (is_above_lower_limit && is_below_hard_stop_limit)
+  {
+    velocity_scale -=
+        (current_condition_number - lower_singularity_threshold) / (upper_threshold - lower_singularity_threshold);
+
+    servo_status = moving_towards_singularity ? StatusCode::DECELERATE_FOR_APPROACHING_SINGULARITY :
+                                                StatusCode::DECELERATE_FOR_LEAVING_SINGULARITY;
+  }
+  // If condition number has crossed hard stop limit, halt the robot.
+  else if (!is_below_hard_stop_limit)
+  {
+    servo_status = StatusCode::HALT_FOR_SINGULARITY;
+    velocity_scale = 0.0;
+  }
+
+  return std::make_pair(velocity_scale, servo_status);
+}
+
+double velocityScalingFactor(const Eigen::VectorXd& velocities, const moveit::core::JointBoundsVector& joint_bounds,
+                             double scaling_override)
+{
+  // If override value is close to zero, user is not overriding the scaling
+  if (scaling_override < 0.01)
```

Comment:
Can we make this 0.01 a constant that is more visible, rather than hard-coded? I've had to explain this one to a lot of people haha

Reply:
```c++
namespace
{
// The threshold above which `override_velocity_scaling_factor` will be used instead of computing the scaling from joint bounds.
const double SCALING_OVERRIDE_THRESHOLD = 0.01;
}  // namespace

```

This should be enough ?